R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+#'   year = c(2020, 2025, 2030),
+#'   plan_alloc_wt_tech_prod = rep(1, 3),
+#'   scen_alloc_wt_tech_prod = 1:3,
+#'   plan_carsten = rep(0.5, 3)
+#' )
+#' pacta_results
+#'
+#' calculate_tdm(pacta_results, t0 = 2020)
+calculate_tdm <- function(data, t0, t1 = 5, t2 = 10, additional_groups = NULL) {
+  stopifnot(
+    is.data.frame(data),
+    is.numeric(t0),
+    is.numeric(t1),
+    is.numeric(t2)
+  )
+
+  if (!is.null(additional_groups)) stopifnot(is.character(additional_groups))
+
+  groups <- union(crucial_tdm_groups(), additional_groups)
+
+  check_calculate_tdm(data, t0, t1, t2, groups)
+
+  filtered_data <- filter(data, .data$allocation == "portfolio_weight")
+
+  if (nrow(filtered_data) == 0) {
+    return(warn_zero_rows(tdm_prototype()))
+  }
+
+  # TODO: Filter the data fora  particular scenario. Function should gain the
+  # argument `scenario`, which by default specifies IPR - FPS scenario. I am
+  # waiting until the data is prepared as I'm not sure exactly what this
+  # scenario will be called in the input data.
+
+  data_with_monotonic_factors <- add_monotonic_factor(filtered_data, t0, t1, t2, groups)
+
+  initial_year_data <- filter(data_with_monotonic_factors, .data$year == t0)
+
+  preformatted_data <- pre_format_data(data_with_monotonic_factors, t0, t1, t2, groups)
+
+  data_with_tdm <- add_tdm(preformatted_data, groups)
+
+  formatted_data_with_tdm <- left_join(
+    initial_year_data,
+    data_with_tdm,
+    by = groups
+  )
+
+  formatted_data_with_tdm %>%
+    add_aggregate_tdm(t0, groups) %>%
+    select(names(tdm_prototype()), all_of(groups))
+}
+
+warn_zero_rows <- function(data) {
+  # NOTE: This function will only work if the allocation method is
+  # portfolio_weight. ownership_weight outputs 0 for all carsten metric values,
+  # and thus we would be dividing by zero otherwise...
+  message <- 'Filtering for "portfolio_weight" allocation, outputs 0 rows'
+  warn(message, class = "has_zero_rows")
+
+  # Pass `data` to inline inside return()
+  invisible(data)
+}
+
+add_time_step <- function(data, t0, t1, t2) {
+  data %>%
+    mutate(
+      time_step = case_when(
+        .data$year == t0 ~ "t0",
+        .data$year == t0 + t1 ~ "plus_t1",
+        .data$year == t0 + t2 ~ "plus_t2"
+      )
+    )
+}
+
+add_monotonic_factor <- function(data, t0, t1, t2, groups) {
+  data <- data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    mutate(
+      end_year = last(.data$year),
+      end_year_is_t0_t2 = .data$end_year == t0 + t2
+    )
+
+  if (all(data$end_year_is_t0_t2)) {
+    monotonic_factors <- data %>%
+      mutate(
+        .is_monotonic = TRUE,
+        monotonic_factor = dplyr::if_else(.data$.is_monotonic, 1, -1),
+        .is_monotonic = NULL
+      ) %>%
+      select(
+        -c(
+          .data$year,
+          .data$end_year,
+          .data$plan_carsten,
+          .data$plan_alloc_wt_tech_prod,
+          .data$scen_alloc_wt_tech_prod,
+          .data$end_year_is_t0_t2
+        )
```

Comment:
is this normal to wrap variable names in a `c()` in `select()`?

Reply:
I have done it frequently when dropping variables (mainly because you can extract the `-` sign out, so it's a little easier to read). I don't know if that means it is `norma;` or not